tests/gstreamer_mqtt/unittest_mqtt_w_helper.cc

```diff
@@ -0,0 +1,886 @@
+  hdr->duration = duration;
+}
+
+/**
+ * @brief Test for mqttsink with GstMqttTestHelper (push a GstBuffer)
+ */
+TEST (testMqttSink, sink_push0)
+{
+  GstHarness *h = gst_harness_new ("mqttsink");
+  GstFlowReturn ret;
+
+  g_object_set (h->element, "debug", true, NULL);
+  gst_harness_add_src_parse (h, "videotestsrc is-live=1 ! queue", TRUE);
+  GstMqttTestHelper::getInstance ().initFailFlags ();
+  ret = gst_harness_push_from_src (h);
+
+  EXPECT_EQ (ret, GST_FLOW_OK);
+
+  gst_harness_teardown (h);
+}
+
+/**
+ * @brief Test for mqttsink with GstMqttTestHelper (push a GstBuffer)
+ */
+TEST (testMqttSink, sink_push1)
+{
+  GstHarness *h = gst_harness_new ("mqttsink");
+  GstFlowReturn ret;
+
+  g_object_set (h->element, "debug", true, NULL);
+  gst_harness_add_src_parse (h, "videotestsrc is-live=1 ! queue", TRUE);
+  GstMqttTestHelper::getInstance ().initFailFlags ();
+  ret = gst_harness_push_from_src (h);
+
+  EXPECT_EQ (ret, GST_FLOW_OK);
+
+  gst_harness_teardown (h);
+}
+
+/**
+ * @brief Test for mqttsink with GstMqttTestHelper (Push multiple GstBuffers with num-buffers)
+ */
+TEST (testMqttSink, sink_push2)
+{
+  GstHarness *h = gst_harness_new ("mqttsink");
+  GstFlowReturn ret;
+  const gint num_buffers = 10;
+  gint i;
+
+  g_object_set (h->element, "num-buffers", num_buffers, NULL);
+  g_object_set (h->element, "debug", true, NULL);
+
+  gst_harness_add_src_parse (h, "videotestsrc is-live=1 ! queue", TRUE);
+  GstMqttTestHelper::getInstance ().initFailFlags ();
+  for (i = 0; i < num_buffers; ++i) {
+    ret = gst_harness_push_from_src (h);
+    EXPECT_EQ (ret, GST_FLOW_OK);
+  }
+
+  gst_harness_teardown (h);
+}
+
+/**
+ * @brief Test for mqttsink with GstMqttTestHelper (MQTTAsync_send failure case)
+ */
+TEST (testMqttSink, sink_push0_n)
+{
+  const static gsize data_size = 1024;
+  GstHarness *h = gst_harness_new ("mqttsink");
+  GstBuffer *in_buf;
+  GstFlowReturn ret;
+
+  ASSERT_TRUE (h != NULL);
+
+  g_object_set (h->element, "debug", true, NULL);
+
+  in_buf = gst_harness_create_buffer (h, data_size);
+  GstMqttTestHelper::getInstance ().initFailFlags ();
+  GstMqttTestHelper::getInstance ().setFailSend (true);
+  ret = gst_harness_push (h, in_buf);
+
+  EXPECT_EQ (ret, GST_FLOW_ERROR);
+  GstMqttTestHelper::getInstance ().setFailSend (false);
+
+  gst_harness_teardown (h);
+}
+
+/**
+ * @brief Test for mqttsink with GstMqttTestHelper (MQTTAsync_disconnect failure case)
+ */
+TEST (testMqttSink, sink_push1_n)
+{
+  const static gsize data_size = 1024;
+  GstHarness *h = gst_harness_new ("mqttsink");
+  GstBuffer *in_buf;
+  GstFlowReturn ret;
+
+  ASSERT_TRUE (h != NULL);
+
+  g_object_set (h->element, "debug", true, NULL);
+  GstMqttTestHelper::getInstance ().initFailFlags ();
+
+  in_buf = gst_harness_create_buffer (h, data_size);
+  GstMqttTestHelper::getInstance ().setFailDisconnect (true);
+  ret = gst_harness_push (h, in_buf);
+
+  EXPECT_EQ (ret, GST_FLOW_OK);
```

Comment:
not `GST_FLOW_ERROR`?

Reply:
@gichan-jang ```GstMqttTestHelper::getInstance ().setFailDisconnect (true);``` invokes the callback handling the failure of  ```MQTTAsync_disconnect()```, which is not related to ```gst_harness_push()```'s return value.

Reply:
Then how do you check it is disconnected?

Reply:
https://github.com/nnstreamer/nnstreamer/blob/9853457eef5968a9658d638ab911789f974a293e/gst/mqtt/mqttsink.c#L1124 will be invoked for gentle finalization of the pipeline.